## Project 3 - Part A

### 1. 
Group Name : Violet
Members : Yifu Deng, Jiangqiu Shen, Zhiyuan Lu

### 2-a
Here we suppose that the root node is {N} and the nodes in two way splits are {N1-1, N1-2} and {N2-1, N2-2, N2-3} repectively.

In [12]:
print('GINI Index for each node')
print('GINI(N) = 1 - (100/210)^2 - (50/210)^2 - (60/210)^2')
gini_n = 1 - (100/210)**2 - (50/210)**2 - (60/210)**2
print('=',gini_n)
print('GINI(N1-1) = 1 - (56/68)^2 - (12/68)^2')
gini_n1_1 = 1 - (56/68)**2 - (12/68)**2
print('=',gini_n1_1)
print('GINI(N1-2) = 1 - (44/142)^2 - (38/142)^2 - (60/142)^2')
gini_n1_2 = 1 - (44/142)**2 - (38/142)**2 - (60/142)**2
print('=',gini_n1_2)
print('GINI(N2-1) = 1 - (62/80)^2 - (18/80)^2')
gini_n2_1 = 1 - (62/80)**2 - (18/80)**2 - 0
print('=', gini_n2_1)
print('GINI(N2-2) = 1 - (28/63)^2 - (11/63)^2 - (24/63)^2')
gini_n2_2 = 1 - (28/63)**2 - (11/63)**2 - (24/63)**2
print('=',gini_n2_2)
print('GINI(N2-3) = 1 - (10/67)^2 - (21/67)^2 - (36/67)^2')
gini_n2_3 = 1 - (10/67)**2 - (21/67)**2 - (36/67)**2
print('=',gini_n2_3)
print('\n\n')
gain_1 = gini_n - (2*gini_n1_1+3*gini_n1_2)/(2+3)
print('The gain in GINI Index for the first splits',gain_1)
gain_2 = gini_n - (2*gini_n2_1+3*gini_n2_2+3*gini_n2_3)/(2+3+3)
print('The gain in GINI Index for the second splits',gain_2)

GINI Index for each node
GINI(N) = 1 - (100/210)^2 - (50/210)^2 - (60/210)^2
= 0.634920634920635
GINI(N1-1) = 1 - (56/68)^2 - (12/68)^2
= 0.2906574394463668
GINI(N1-2) = 1 - (44/142)^2 - (38/142)^2 - (60/142)^2
= 0.6538385241023607
GINI(N2-1) = 1 - (62/80)^2 - (18/80)^2
= 0.3487499999999999
GINI(N2-2) = 1 - (28/63)^2 - (11/63)^2 - (24/63)^2
= 0.6268581506676745
GINI(N2-3) = 1 - (10/67)^2 - (21/67)^2 - (36/67)^2
= 0.5907774560035643



The gain in GINI Index for the first splits 0.12635454468067187
The gain in GINI Index for the second splits 0.09111978241892038


### 2-b
Based on selecting node with largest gain in GINI index, N1-1 and N1-2 are preferred.

### 2-C

In [42]:
import math
print('H(N) = - {P(Y=A)logP(Y=A) + P(Y=B)logP(Y=B) +P(Y=C)logP(Y=C)}')
h_n = -( (100/210)*math.log2(100/210) + (50/210)*math.log2(50/210) + (60/210)*math.log2(60/210) )
print('Entropy before split is H(N)=', h_n)

print('\nFor the first split 1,')
p1 = 68/210
p2 = 142/210
print('P(X = N1-1) is ', p1,'  P(X=N1-2) is ', p2) 
pA_1 = 56/68
pB_1 = 12/68
pC_1 = 0/68
print('P(Y=A|N1-1) is ',pA_1,'  P(Y=B|N1-1) is ',pB_1,'  P(Y=C|N1-1) is ',pC_1)
pA_2 = 44/142
pB_2 = 38/142
pC_2 = 60/142
print('P(Y=A|N1-2) is ',pA_2,'  P(Y=B|N1-2) is ',pB_2,'  P(Y=C|N1-2) is ',pC_2)
h_s1 = - ( p1*( pA_1*math.log2(pA_1)+pB_1*math.log2(pB_1)) + p2*( pA_2*math.log2(pA_2)+pB_2*math.log2(pB_2)+pC_2*math.log2(pC_2) ) )
print('Entropy after split is H(N|S1) =',h_s1, '.  Information gain is H(N)-H(N|S1) =', h_n-h_s1)

print('\nFor the split 2,')
_p1 = 80/210
_p2 = 63/210
_p3 = 67/210
print('P(X = N2-1) is ', _p1,'  P(X=N2-2) is ', _p2, '  P(X=N2-3) is ', _p3)
_pA_1 = 62/80
_pB_1 = 18/80
_pC_1 = 0/80
print('P(Y=A|N2-1) is ',_pA_1,'  P(Y=B|N2-1) is ',_pB_1,'  P(Y=C|N2-1) is ',_pC_1)
_pA_2 = 28/63
_pB_2 = 11/63
_pC_2 = 24/63
print('P(Y=A|N2-2) is ',_pA_2,'  P(Y=B|N2-2) is ',_pB_2,'  P(Y=C|N2-2) is ',_pC_2)
_pA_3 = 10/67
_pB_3 = 21/67
_pC_3 = 36/67
print('P(Y=A|N2-3) is ',_pA_3,'  P(Y=B|N2-3) is ',_pB_3,'  P(Y=C|N2-3) is ',_pC_3)
h_s2 =0
h_s2 += _p1*(_pA_1 * math.log2(_pA_1) + _pB_1*math.log2(_pB_1) )
h_s2 += _p2*( _pA_2*math.log2(_pA_2) + _pB_2*math.log2(_pB_2) + _pC_2*math.log2(_pC_2) ) 
h_s2 += _p3*( _pA_3*math.log2(_pA_3) + _pB_3*math.log2(_pB_3) + _pC_3*math.log2(_pC_3) )
h_s2 = -h_s2
print('Entropy after split is H(N|S2) =',h_s2, '.  Information gain is H(N)-H(N|S2) =', h_n - h_s2)


H(N) = - {P(Y=A)logP(Y=A) + P(Y=B)logP(Y=B) +P(Y=C)logP(Y=C)}
Entropy before split is H(N)= 1.5190461643198376

For the first split 1,
P(X = N1-1) is  0.3238095238095238   P(X=N1-2) is  0.6761904761904762
P(Y=A|N1-1) is  0.8235294117647058   P(Y=B|N1-1) is  0.17647058823529413   P(Y=C|N1-1) is  0.0
P(Y=A|N1-2) is  0.30985915492957744   P(Y=B|N1-2) is  0.2676056338028169   P(Y=C|N1-2) is  0.4225352112676056
Entropy after split is H(N|S1) = 1.2710974574447609 .  Information gain is H(N)-H(N|S1) = 0.24794870687507675

For the split 2,
P(X = N2-1) is  0.38095238095238093   P(X=N2-2) is  0.3   P(X=N2-3) is  0.319047619047619
P(Y=A|N2-1) is  0.775   P(Y=B|N2-1) is  0.225   P(Y=C|N2-1) is  0.0
P(Y=A|N2-2) is  0.4444444444444444   P(Y=B|N2-2) is  0.1746031746031746   P(Y=C|N2-2) is  0.38095238095238093
P(Y=A|N2-3) is  0.14925373134328357   P(Y=B|N2-3) is  0.31343283582089554   P(Y=C|N2-3) is  0.5373134328358209
1.1917047844910704
Entropy after split is H(N|S2) = 1.1917047844910704 .  Informati

### 2-d
Based on selecting variable to maxmize information gain, the second way to split is preferred.

### 3-a
Fruit Data

Columns: Type, Weight, Height, Width

Type: 
 1 - apples
 2 - oranges
 3 - lemons

Weight (grams)
  0  - if wt <= 179.42 
  1  - otherwise

Height (cm)
  2 - if ht > 8.5
  1 - if ht <= 8.5 && > 7.3
  0 - otherwise

Width (cm)
  2 - if width > 7.8
  1 - if width <= 7.8 && >7.3
  0 - otherwise



In [77]:
import pandas as pd
data = pd.read_csv("./data/fruit.txt",names=['T','Wei','H','Wid'])
data = data.loc[data['T']<3]
#print(data)
appl = data.loc[data['T']==1]
orag = data.loc[data['T']==2]
p_appl = len(appl)/len(data)
p_orag = len(orag)/len(data)
print(p_appl,p_orag) 

p_wt0_a  = len(appl.loc[appl['Wei']==0])/len(appl)
p_wt1_a  = len(appl.loc[appl['Wei']==1])/len(appl)
p_wt0_o  = len(orag.loc[orag['Wei']==0])/len(orag)
p_wt1_o  = len(orag.loc[orag['Wei']==1])/len(orag)
print(p_wt0_a,p_wt1_a,p_wt0_o,p_wt1_o)

p_h0_a  = len(appl.loc[appl['H']==0])/len(appl)
p_h1_a  = len(appl.loc[appl['H']==1])/len(appl)
p_h2_a  = len(appl.loc[appl['H']==2])/len(appl)
p_h0_o  = len(orag.loc[orag['H']==0])/len(orag)
p_h1_o  = len(orag.loc[orag['H']==1])/len(orag)
p_h2_o  = len(orag.loc[orag['H']==2])/len(orag)
print(p_h0_a,p_h1_a,p_h2_a ,p_h0_o ,p_h1_o ,p_h2_o)

p_wid0_a  = len(appl.loc[appl['Wid']==0])/len(appl)
p_wid1_a  = len(appl.loc[appl['Wid']==1])/len(appl)
p_wid2_a  = len(appl.loc[appl['Wid']==2])/len(appl)
p_wid0_o  = len(orag.loc[orag['Wid']==0])/len(orag)
p_wid1_o  = len(orag.loc[orag['Wid']==1])/len(orag)
p_wid2_o  = len(orag.loc[orag['Wid']==2])/len(orag)
print(p_wid0_a,p_wid1_a,p_wid2_a ,p_wid0_o ,p_wid1_o ,p_wid2_o)

0.5 0.5
0.8947368421052632 0.10526315789473684 0.631578947368421 0.3684210526315789
0.3157894736842105 0.6842105263157895 0.0 0.5789473684210527 0.2631578947368421 0.15789473684210525
0.5789473684210527 0.3684210526315789 0.05263157894736842 0.21052631578947367 0.3684210526315789 0.42105263157894735


;|Prior|Prob|
;| - | - |
;|P(apple)| 0.5|
;|P(orange)|0.5|

;|Cond.| Prob.|
;| - | - |
;|P(Wt = 0/apple) | 0.8947368421052632 |
;|P(Wt = 1/apple) | 0.10526315789473684|
;|P(Wt = 0/orange) |  0.631578947368421 |
;|P(Wt = 1/orange)| 0.3684210526315789|


;|Cond.| Prob.|
;| - | - |
;|P(Ht = 0/apple) |0.3157894736842105 |
;|P(Ht = 1/apple) |  0.6842105263157895 |
;|P(Ht = 2/apple) | 0.0 |
;|P(Ht = 0/orange) |0.5789473684210527 |
;|P(Ht = 1/orange) | 0.2631578947368421 |
;|P(Ht = 2/orange) | 0.15789473684210525|

   

|Cond.| Prob.|
| - | - |
|P(Wid = 0/apple) | 0.5789473684210527 |
|P(Wid = 1/apple)  |  0.3684210526315789 | 
|P(Wid = 2/apple)  | 0.05263157894736842 |
|P(Wid = 0/orange) | 0.21052631578947367 |
|P(Wid = 1/orange) | 0.3684210526315789 |
|P(Wid = 2/orange)|  0.42105263157894735 |